### DISCUSSION TITLE
Latch with a three-state input

### DISCUSSION THREAD

**Discussion by jnoche** [hneemann/Digital#1174] [Q&A] [answered]
I need a latch (whose output Q is initially 0) that has a three-state input D and if that is at a high impedance, then the output is unchanged.  For example, if the input D goes 0, Z, 1, Z, 0, 1, Z, 1, 0, Z, then the output should go 0, 0, 1, 1, 0, 1, 1, 1, 0, 0.  My attempt is shown below.

![D](https://github.com/hneemann/Digital/assets/63948655/9193b83a-ddab-4630-8b72-0e0866473d0e)

But this doesn't work as I want it.  What happens is if the input is Z, then the output is unpredictable.  I tried using the built-in RS latch (see below)

![RS](https://github.com/hneemann/Digital/assets/63948655/bd6b62ac-7892-4b58-869e-7ba4b2c58034)

but it also doesn't work as I want it.  (A high-impedance input causes an unpredictable output.)

Is there a way to create the component that I want using the Digital software?

**Comment by mengstr** [comment]
I think this contraption does what you want... maybe.... ;-)

<img width="679" alt="hold3state" src="https://github.com/hneemann/Digital/assets/325326/db00c362-0359-4fab-9ba1-9eeb55b9182e">

**Reply by hneemann** [reply]
This circuit is interesting, but feels to me like a misuse of the splitters. My variant feels better, but the more I think about it, the smaller the difference becomes, because my variant starts to feel like a misuse of the FETs. :smile:

**Reply by mengstr** [reply]
:-) I was first thinking of using "drivers" but I couldn't figure out a nice way of using them. My mind then went to transmission gates and the fets, but I never actually used neither of them in Digital. But I remembered that the splitters are (annoyingly) one-way devices so they would be perfect to isolate two parts of a circuit while still being able to transmit data.

You design looks more what a "real" circuit could be designed as. Mine is a design possible by a Digital quirk.

IRL one would of course make a discrete bus-hold with a buffer and a resistor or just use a premade 74-series bus-hold driver/gate.
<img width="534" alt="bushold" src="https://github.com/hneemann/Digital/assets/325326/e79cf4c6-2ca9-4b77-aac0-84a0dd271eee">

**Reply by jnoche** [reply]
Thank you very much for this.  I did not know that the splitter can "isolate" the input from the output.  Here, it allows the input to be at high impedance while the output can be either pulled up or pulled down.

**Reply by jnoche** [reply]
I hope you don't mind. I'll be marking the other answer as "the" answer because, as you said, your circuit seems to work only because of how the Digital software treats splitters.

**Reply by mengstr** [reply]
No worries at all. A universal applicable solution is always the best.

**Reply by hneemann** [reply]
The "Bus-hold" or "pin keeper" is not possible in this simulator because there are no real resistors. But you can build it like this:

![pinKeeper](https://github.com/hneemann/Digital/assets/7190563/a3769fd3-a011-4e14-9b6f-70f0ef55fe0c)

This would possibly be an even more elegant solution to the original question and would also actually be possible to build in reality.

**Reply by mengstr** [reply]
Ah yes, that's definitely the best solution here!

**Reply by jnoche** [reply]
How do I modify this simpler circuit so that its initial value (in Digital) is 0?

**Reply by hneemann** [reply]
Somewhere on the bus a reset logic must be added:

![pinKeeperReset](https://github.com/hneemann/Digital/assets/7190563/e5034ed7-50a3-4b2f-ac3e-f2fb09433380)

**Reply by jnoche** [reply]
Thank you for this.

**Comment by hneemann** [comment, marked as answer]
When a real circuit is built in CMOS technology, an open input is unpredictable and can damage the circuit. And also in TTL an open input is not advisable, although it is recognized as high. The simulator simulates this "unpredictability" by generating random values.
But you can detect high-z in the simulator with a special circuit, and build an according logic.

![high-z-rs-ff](https://github.com/hneemann/Digital/assets/7190563/598b3243-06a5-4268-bce8-deb72583dc7f)

**Reply by jnoche** [reply]
Thank you very much for this.  Is there any particular reason the transistor at the top is floating gate and the one at the bottom is not?

**Reply by hneemann** [reply]
>  Is there any particular reason the transistor at the top is floating gate and the one at the bottom is not?

Thanks for the hint! No, there is no reason! I made a mistake, which I have now corrected!

**Comment by jnoche** [comment]
I'm writing this information here so that future readers can benefit from it.  I've decided to use the Neemann circuit with the RS latch (which I'll call Neemann1) instead of the simpler Neemann circuit (Neemann2).  It's because the simpler circuit results in an error when it is placed in the complex design I have.

Neemann1 with D=Z looks like this:
![Neemann1](https://github.com/hneemann/Digital/assets/63948655/7181b017-b904-4463-8cdd-7c864fdc6b76)
Note that the input box is gray with a white background; it remains at high impedance.

Neemann2 with D=Z looks like this:
![Neemann2](https://github.com/hneemann/Digital/assets/63948655/9ca77642-8bc5-4e3f-9b7c-f8eeee0ec0da)
Note that the input box is gray with a light green background; it used to be high impedance but is now pulled up.  The external circuit connected to the input is affected, causing an error.
